### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] false error when type parameter is used only in assertion function

### ISSUE THREAD

**Issue by jdufresne** [typescript-eslint/typescript-eslint#9895]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=GYVwdgxgLglg9mABMAPAFUQUwB5U2AEwGdE4AjAK02gD4AKbALlMuqgEpmBDIozAJyglsiGCQwBvAFCJE2ANxSAvlKk4ADnEGICmYFxAAbKMkVA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwp4BXWkQxhQEMdnyES5SjUT0mzBgHsqfBg2TEiiBHiqTkVNrAC2yHKkTDIqeMrQ8AviCdA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function f<T extends object>(x: object): asserts x is T {
  x;
}

export default f;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No errors as the parameter T is used in the assertion `asserts x is T`. The function is parameterized as the type T is different at different call sites.

### Actual Result

```
Type parameter T is used only once in the function signature. 1:12 - 1:28
```

**Comment by JoshuaKGoldberg** [comment]
But, type parameter `T` _is_ used only once in the function signature. There's nothing stopping users of `f` from passing their own type parameter:

```ts
function f<T extends object>(x: object): asserts x is T {
  x;
}

const box = {};

f<{ value: string }>(box);

box.value.toUpperCase(); // Should type error, but doesn't
```

I'd think this is working as intended, no? Do you have an example where this isn't?

**Comment by jdufresne** [comment]
I see what you're saying. You're right. I misunderstood how this scenario was expected to work. Thanks for the response and I'll close the issue.
